perceval/components/source.py

```diff
@@ -21,66 +21,123 @@
+        px = self.occupation_factor
+
+        # g2 = 2p2/(p1+2p2)**2
+        # p1 + p2 = px
+
+        p2 = min(np.poly1d([g2, -2 * (1 - g2 * px), g2 * px ** 2]).r)
+        p1 = px - p2
+
         svd = SVDistribution()
-        if self.brightness != 1:
-            svd[StateVector([0])] = 1-self.brightness
+
+        svd[StateVector([0])] = 1-(eta*p1+eta**2*p2+2*eta*(1-eta)*p2)
+
         if self._indistinguishability_model == "homv":
             distinguishability = 1-math.sqrt(self.indistinguishability)
         else:
-            distinguishability = 1 - self.indistinguishability
+            distinguishability = 1-self.indistinguishability
+
         # Approximation distinguishable photons are pure
-        if self.purity != 1:
-            if distinguishability:
-                if self._purity_model == "random":
-                    random_feat = self._context["discernability_tag"]
-                    svd[StateVector([2], {0: ["_:0", "_:%s" % random_feat]})] = self.brightness * (1 - self.purity)
-                    self._context["discernability_tag"] += 1
+        distinguishable_photon = self._context["discernability_tag"]
+        self._context["discernability_tag"] += 1
+        noise_photon = self._context["discernability_tag"]
+        self._context["discernability_tag"] += 1
+
+        if p2 != 0:
+            if distinguishability != 0:
+                if self._multiphoton_model == "distinguishable":
+                    svd[StateVector([2], {1: {"_": 0}, 2: {"_": noise_photon}})] = eta**2*(1-distinguishability)*p2
+                    svd[StateVector([2], {1: {"_": distinguishable_photon}, 2: {"_": noise_photon}})] = eta**2*distinguishability*p2
                 else:
-                    svd[StateVector([2], {0: ["_:0", "_:0"]})] = self.brightness * (1 - self.purity)
+                    svd[StateVector([2], {1: {"_": 0}, 2: {"_": 0}})] = eta**2*(1-distinguishability)*p2
+                    svd[StateVector([2], {1: {"_": distinguishable_photon}, 2: {"_": 0}})] = eta**2*distinguishability*p2
             else:
-                svd[StateVector([2])] = self.brightness*(1-self.purity)
-        if distinguishability:
-            random_feat = self._context["discernability_tag"]
-            self._context["discernability_tag"] += 1
-            svd[StateVector([1], {0: ["_:%s" % random_feat]})] = distinguishability*self.brightness*self.purity
-            svd[StateVector([1], {0: ["_:0"]})] = (1-distinguishability)*self.brightness*self.purity
+                if self._multiphoton_model == "distinguishable":
+                    svd[StateVector([2], {1: {"_": 0}, 2: {"_": noise_photon}})] = eta**2*p2
+                else:
+                    svd[StateVector([2])] = eta**2*p2
+
+        if distinguishability != 0:
+            if self._multiphoton_model == "distinguishable":
+                svd[StateVector([1], {1: {"_": distinguishable_photon}})] = eta*distinguishability*p1+eta*(1-eta)*distinguishability*p2
```

Comment:
l.126 and l.130 are the same. You can put it outside the condition scope

Reply:
self._multiphoton_model can be either "indistinguishable" or "distinguishable"